File: TF_Service_dotNet/TouchFree.Tooling/InputActionPlugin.cs
using Ultraleap.TouchFree.Library;

namespace Ultraleap.TouchFree.Tooling
{
    // Class: InputActionPlugin
    // A base class for Plugins which are used by the <InputActionManager> to manipulate the
    // <InputAction> data received from the Service.
    public abstract class InputActionPlugin
    {
        // Variable: InputActionOutput
        // An event for transmitting <InputActions> as they pass through this plugin.
        // This can be used to access the data as it is used by a specific plugin, as to intercept
        // the full cycle of plugins that the <InputActionManager> references.
        public event InputActionManager.InputActionEvent InputActionOutput;

        // Function: RunPlugin
        // Called from <InputActionManager> and provided a <InputAction> as a parameter.
        // This function is a wrapper that guarantees that the results of <ModifyInputAction> are both
        // returned to the <InputActionManager> and transmitted via <TransmitInputAction>.
        public ClientInputAction? RunPlugin(ClientInputAction _inputAction)
        {
            ClientInputAction? modifiedInputAction = ModifyInputAction(_inputAction);

            if (modifiedInputAction.HasValue)
            {
                TransmitInputAction(modifiedInputAction.Value);
            }

            return modifiedInputAction;
        }

        // Function: ModifyInputAction
        // Called from <RunPlugin> and provided a <InputAction> as a parameter.
        // This function is used to manipulate the incoming <InputAction>
        // data. Returns a <InputAction> which is then distributed via the <InputActionManager>.
        protected abstract ClientInputAction? ModifyInputAction(ClientInputAction _inputAction);

        // Function: TransmitInputAction
        // To be used to Invoke the <InputActionOutput> event from any child class of this base.
        internal void TransmitInputAction(ClientInputAction _inputAction)
        {
            InputActionOutput?.Invoke(_inputAction);
        }
    }
}

File: TF_Tooling_Web/src/Plugins/InputActionPlugin.ts
import { TouchFreeInputAction } from '../TouchFreeToolingTypes';

/**
 * Base class for input action plugins
 * @remarks
 * The `InputActionManager` runs each plugin upon receiving a message
 * from the service before dispatching an InputAction event.
 * Input action plugins invoke a `"InputActionOutput"` event on themselves
 * for subscribers to listen to if the results of a specific plugin is required.
 * @public
 */
export abstract class InputActionPlugin extends EventTarget {
    /**
     * Run this plugin, modifying the `InputAction` and dispatching an `"InputActionOutput"` event from this plugin
     * @param _inputAction Input action input
     * @returns Modified input action
     */
    RunPlugin(_inputAction: TouchFreeInputAction): TouchFreeInputAction | null {
        const modifiedInputAction = this.ModifyInputAction(_inputAction);

        if (modifiedInputAction != null) {
            this.TransmitInputAction(modifiedInputAction);
        }

        return modifiedInputAction;
    }

    /**
     * Proxy function for derived classes to modify input actions before they are dispatched.
     * 
     * @internal
     */
    ModifyInputAction(_inputAction: TouchFreeInputAction): TouchFreeInputAction | null {
        return _inputAction;
    }

    /**
     * For derived classes to invoke the `InputActionOutput` event.
     * @param _inputAction InputAction state to dispatch event with
     * 
     * @internal
     */
    TransmitInputAction(_inputAction: TouchFreeInputAction): void {
        const InputActionEvent: CustomEvent<TouchFreeInputAction> = new CustomEvent<TouchFreeInputAction>(
            'InputActionOutput',
            { detail: _inputAction }
        );
        this.dispatchEvent(InputActionEvent);
    }
}
